pyccel/parser/syntactic.py

```diff
@@ -238,7 +240,60 @@ def _visit(self, stmt):
 
 
     def _visit_RedBaron(self, stmt):
-        code = CodeBlock([self._visit(i) for i in stmt])
+        prog = []
+        mod  = []
+        start = []
+        current_file = start
+        targets = []
+        n_empty_lines = 0
+        is_prog = False
+        for i in stmt:
+            v = self._visit(i)
+            if n_empty_lines > 3:
+                current_file = start
+            if isinstance(v,(FunctionDef, ClassDef)):
+                n_empty_lines = 0
+                mod.append(v)
+                targets.append(v.name)
+                current_file = mod
+            elif isinstance(v,(Header,Comment)):
+                n_empty_lines = 0
+                current_file = start
+                current_file.append(v)
+            elif isinstance(v, (NewLine, EmptyLine)):
+                current_file.append(v)
+                n_empty_lines += 1
+            elif isinstance(v, Import):
+                n_empty_lines = 0
+                mod.append(v)
+                prog.append(v)
+            else:
+                is_prog = True
+                n_empty_lines = 0
+                prog.append(v)
+                current_file = prog
+            if len(start)>0 and current_file is not start:
+                current_file[-1:-1] = start
+                start = []
+        current_mod_name = os.path.splitext(os.path.basename(self._filename))[0]
+        prog_name = 'prog_' + current_mod_name
+        mod_code = CodeBlock(mod) if len(targets)>0 else None
+        if is_prog:
+            if mod_code:
+                expr = Import(targets, source=current_mod_name)
+                prog.insert(0,expr)
+            prog_code = CodeBlock(prog)
+            prog_code.set_fst(stmt)
+        else:
+            prog_code = None
+            if mod_code is None:
+                mod_code = CodeBlock(mod+start)
+        assert( mod_code is not None or prog_code is not None)
+        from pyccel.ast import ParserResult
+        code = ParserResult(program   = prog_code,
+                            module    = mod_code,
+                            prog_name = prog_name,
+                            mod_name  = current_mod_name)
         code.set_fst(stmt)
         return code
```

Comment:
Can you explain the use of these additional 50 lines of code?

Reply:
They break the syntax ast into 2 CodeBlocks. One for the module and one for the Program. FunctionDefs and ClassDefs always go into modules, Headers and Comments go in the same CodeBlock as the following object, EmptyLines go in the same CodeBlock as the previous line. Imports go in both CodeBlocks (this means that there can be unnecessary imports), everything else goes in the program

Reply:
Thanks for the explanation. Could you copy and paste it as a comment in the source code?